### PR TITLE
ledger: avoid producing empty local delta

### PR DESCRIPTION
**Pull request by algorandskiy**
## Summary

Ensure local delta does not have entries with empty key-value stores

## Test Plan

New test added

### CODE REVIEW

File: ledger/appcow.go [hunk 1]
```diff
@@ -467,7 +467,12 @@ func (cb *roundCowState) BuildEvalDelta(aidx basics.AppIndex, txn *transactions.
 				if err != nil {
 					return basics.EvalDelta{}, err
 				}
-				evalDelta.LocalDeltas[addrOffset] = sdelta.kvCow.serialize()
+
+				d := sdelta.kvCow.serialize()
+				// ignore empty deltas
+				if len(d) != 0 {
```
Comment (tsachiherman) [inline ledger/appcow.go:473]: while it's probably correct, I think that you ignored the case where `len(d) == 0` and `len(evalDelta.LocalDeltas[addrOffset]) != 0`.
i.e. you would need to add a `delete(evalDelta.LocalDeltas, addrOffset)` here. I'm not sure how this would work in the previous release, and it could be a  non-issue here as well.

File: ledger/appcow_test.go [hunk 2]
```diff
@@ -413,7 +413,7 @@ func TestCowBuildDelta(t *testing.T) {
 	a.Equal(
 		basics.EvalDelta{
 			GlobalDelta: basics.StateDelta{},
-			LocalDeltas: map[uint64]basics.StateDelta{0: {}},
+			LocalDeltas: map[uint64]basics.StateDelta{},
```
Comment (winder) [inline ledger/appcow_test.go:416]: Is there another test case needed for empty sender delta + non-empty other account deltas?

File: ledger/appcow.go (hunk 1 above)
Comment (algorandskiy) [reply ledger/appcow.go:473]: 1. all txn have independent eval delta
2. there is only one kv per addr+appId combination
3. I'll add couple more tests to document this behavior

File: ledger/appcow_test.go (hunk 2 above)
Comment (algorandskiy) [reply ledger/appcow_test.go:416]: Not really needed, but added this test to document this scenario

**Comment by algorandskiy** [comment]
Added couple more tests per review

File: config/consensus.go [hunk 3]
```diff
@@ -872,9 +875,25 @@ func initConsensusProtocols() {
...
+	v27.NoEmptyLocalDeltas = true
+
+	Consensus[protocol.ConsensusV27] = v27
+
+	// v26 or v25 can be upgraded to v27, with an update delay of 3 days
+	// 60279 = (3 * 24 * 60 * 60 / 4.3)
+	// for the sake of future manual calculations, we'll round that down
+	// a bit :
+	v26.ApprovedUpgrades[protocol.ConsensusV27] = 60000
+	v25.ApprovedUpgrades[protocol.ConsensusV27] = 60000
```
Comment (tsachiherman) [inline config/consensus.go:892]: for simplicity, let's limit this to v26 ? The v25 was on betanet only anyway.

File: ledger/appcow.go [hunk 4]
```diff
@@ -467,7 +467,11 @@ func (cb *roundCowState) BuildEvalDelta(aidx basics.AppIndex, txn *transactions.
 				if err != nil {
 					return basics.EvalDelta{}, err
 				}
-				evalDelta.LocalDeltas[addrOffset] = sdelta.kvCow.serialize()
+
+				d := sdelta.kvCow.serialize()
+				if !cb.proto.NoEmptyLocalDeltas || len(d) != 0 {
+					evalDelta.LocalDeltas[addrOffset] = d
+				}
```
Comment (tsachiherman) [inline ledger/appcow.go:474]: this part need to be contingent upon the consensus parameter that you've added.

**Review by tsachiherman — CHANGES_REQUESTED** [review]
only one comment. I don't think we need this line:
```go
v25.ApprovedUpgrades[protocol.ConsensusV27] = 60000
```

otherwise, I think that it looks good.

File: config/consensus.go (hunk 3 above)
Comment (algorandskiy) [reply config/consensus.go:892]: Fixed

**Review by tsachiherman — APPROVED** [review]
looks good to me.

**Review by tsachiherman — APPROVED** [review]
